### ISSUE TITLE
RFC: `@rx-angular/state` - new `rxState`

### ISSUE THREAD

**Issue by hoebbelsB** [rx-angular/rx-angular#1526]
# Motivation

> All changes are non breaking

On top of the general motivations stated in #1592 a couple of things are worth to mention:

- As signals will land in angular 16, we want to provide first class support for the new reactive primitive provided by the
framework and integrate them smoothly with our existing state management solution. The goal is to be able to combine the power of `Observable`s and `Signal`s and integrate them seamlessly to have the best of both worlds.
- As we have now a onDestroy hook (`DestroyRef`) we no longer need to rely on the class based architecture at all and use the inject functions only.
- We can reconsider the old shorthand (`RxState#hold`) for a direct connection between state and effect. 
- We can include a signal compatible API

We've came up with the following core features that have to be supported (on top of all existing observable based APIs):

* connect accepts `Signal`
* get computed from state
* get single property `Signal` from state
* computed is added as top level API
* create `Effect` from on state for:
  * selection
  * computed

# Minimal Example

```typescript
import { rxState } from '@rx-angular/state/state';

@Component()
export class AnyComponent {
  state = rxState<{ name: string }>();
}
```

# State

The following code snippets and explanations showcase the APIs as we think they should be implemented in `RxState`.

## Usage in Component

```ts
type ListViewState = {
    list: Movie[],
    query: string
};

type ListViewUi = {
    searchInput: string;
};

@Component({
    selector: 'list-view',
    template: `
      <input name="search" (change)="ui.searchInput($event)" /><br/>
      <ul>
         ...
      </ul>
  `
})
export class ListViewContainerComponent {
   protected ui = rxActions<ListViewUi>();
    private movieState = inject(MovieService);
    private state = rxState<ListViewState>(({connect}) => {
        connect('list', this.movieState.movies);
        connect('query', this.ui.searchInput$.pipe(debounceTime(300)));
    });
}
```

## Usage in a Service

```ts
type MovieStateActions = {
    refresh: void
    save: MoviePostDto
}
type MovieState = {
    movies: Movie[]
}

@Injectable({
    providedIn: 'root'
})
export class MovieService {
    private moviesResource = inject(MovieResource);
    private state = rxState<MovieState>(({set, connect}) => {
        set({movies: []});
        connect('movies', this.actions.save$,
            ({movies}, updatedMovie) => upsert(movies, updatedMovie, 'id')
        );
        connect(
            'movies',
            this.actions.refresh$.pipe(
                exhaustMap(_ => this.moviesResource.getMovies())
            )
        );
    });
   
    refreshMovies(): void {
        this.actions.refresh();
    }

    updateMovie(movie: Movie): void {
        this.actions.save(movie);
    }
}
```

##  Connect a signal in setup function

We will add another overload to the `connect` method that enables you to populate values coming
from a signal directly into `RxState`.
Internally an `Effect` will be created which listens to to `Signal` and updates the state object accordingly.

It is still possible to connect Observables.

```ts
@Injectable({
    providedIn: 'root'
})
export class MovieService {
 
  private globalState = inject(GlobalState);

  private state = rxState<MovieState>(({connect}) => {
      connect(
           'movies', 
           this.globalState.movies(),
          ({movies}, newMovie) => upsert(movies, newMovie), 'id')
      );
});
```

##  Connect a signal top level API

```ts
@Injectable({
    providedIn: 'root'
})
export class MovieService {
    private globalState = inject(GlobalState);
    private state = rxState<MovieState>();
   
   constructor() {
       this.state.connect(
           'movies', 
           this.globalState.movies(),
          ({movies}, newMovie) => upsert(movies, newMovie), 'id')
        );
   }
}
```

## Computed from state

A convenience function which returns a `computed` based off the properties defined for the state interface.
You can pass a function that receives all available state signals in order to create state derivations.

This is by far the most useful API and highlights the concepts described in the component state reference section.
As `RxState` is completely lazy, signals will be created on-demand - the first time you access them.

```ts
type ListViewState = {
    list: Movie[],
    query: string
};

@Component({
    standalone: true,
    imports: [NgFor],
    selector: 'list-view',
    template: `
      <input name="search" (change)="state.set({query: $event})" /><br/>
      <ul>
        <li *ngFor="let movie of movies()">{{movie.name}}</li>
      </ul>
  `
})
export class ListViewContainerComponent {
    private state = rxState<ListViewState>();

    protected movies = this.state.computed('list');
    // create a computed from state slices, equivalent to `this.state.selectSlice(['list', 'query'], ({ list, query }) => /*..*/)
    protected movies = this.state.computed(({list, query}) => list.filter(m => m.name.includes(query)));
}
```

## State and effects with `onSelect`

```ts
type ListViewState = {
    list: Movie[],
    query: string
};

@Injectable({
    providedIn: 'root'
})
export class TrackingService {
    private http: HttpClient;
    private ef = rxState<>(({onSelect}) => {
        // select slice from 1 prop => this.select('list')
        onSelect('query', ({list, query}) => list.filter(i => i.includes(query)), (m) => this.http.post(m) ));
        // select slice from multiple props => this.select(['list', 'query'])
        onSelect(['list', 'query'], ({list, query}) => list.filter(i => i.includes(query)), (m) => this.http.post(m) ));\
        // select slice with operatorFn
        onSelect(map(({ list }) => mapList(list)), list => this.http.post(list));
    })

}
```

## State and effects with `onComputed`

```ts
type ListViewState = {
    list: Movie[],
    query: string
};

@Injectable({
    providedIn: 'root'
})
export class TrackingService {
    private http: HttpClient;
    private ef = rxState<>(({onComputed}) => {
        onComputed(
            ({list, query}) => list().filter(i => i.includes(query())), 
            (m) => this.http.post(m)
        );
    })

}
```

## Mutate/Set signals

A signal that is returned by `RxState` is not mutable, you can only consume and/or react to its values. 
That being said, state signals will get updated by any method that sets a new value of `RxState`.

```ts
const count: Signal<number> = this.state.computed(({ todos }) => {
    return todos().length;
});

effect(() => {
    console.log(count());
 });

// effect logs: 5

// setting a new value for the 'todos' key will update its matching signal as well
this.state.set({ todos: [] });

// effect logs: 0
```

**Comment by eneajaho** [comment]
This RFC is awesome 🎉

I like all this: 
```ts
private state = inject<RxState<{ todos: Todo[]; foo: string}>>(RxState);

// get signal
const todos: Signal<Todo[]> = this.state.signal('todos');

// set signal
this.state.set({ todos: [] });

// get computed
const count: Signal<number> = this.state.computed(({ todos }) => {
    return todos().length;
});

// connect signal to state
const foo = signal('');
this.state.connect('foo', foo);
```

We currently have `hold` to manage effects, do we really need the new `effect` api, just manage signal effects? 
What does it offer more than Angular own `effect` function except the function that returns the signals?

**Comment by BioPhoton** [comment]
Regards hold vs effect... 
we will rename/deprecate hold.

**Comment by adyngom** [comment]
It definitely is very easy to read and holds an intuitive almost self documenting aspect to it.

**Comment by robert-king** [comment]
looks good,
do we still need to provide RxState, RxEffects and RxAction factory? not sure how weak-refs work, i guess it's nice to be able to provide state at different injector levels depending on how shared the state is between components.

Also the RxActionFactory isn't mentioned, would it be beneficial to have some of these signals convenience methods added to ui.actions? 

Also perhaps there's a way to configure RxAngular to default to signals or observables on a per-component bases, that way i could get compiler errors everywhere were i should change an observable to a signal, but i can also have backwards compatibility if i want. (It looks like you're going for backwards compatibility, although it'd be nice to be able to break things purposefully so that we can update our code to preferred methodology).

**Comment by mhamri** [comment]
I downvote this

1- signal in angular is in a discovery state, which means it may work or it may not work. and it is depending on the team's findings after converting big projects to signals and if everything is still working as it should it's safe to build some API around that. I posted other places and mentioned it to @BioPhoton before. Signal has a memory issue and WeakRef is just one shot at it to see if that gets fixed or not. 
2- solids' signal is already working out of the box with angular; why we should re-implement it? 
3- preview version of angular is coming; what value are we adding here? 
4- **and most importantly,** rx-angular/state is used in production and not everyone is following latest versions of angular, getting dependency on a preview feature stop companies to update to latest version of rx-angular/state to get bug fixes, new features. ... 

In against what we have here, I suggest rx way of signals. I prefer being able to inject multiple RxState and be able to build multiple states around it

Moreover, RxState can have a single-type version that still works with observables

I already created a prototype of this and worked, I park it here as a suggestion

explanation inside the comments
```TS
/**
 to return an observable and at the same time has signal API I used this way
 I'm not sure if attaching more items to an observable is a good idea or not, but for the purpose of 
 maintaining signal API I just bind a function to an observable
 here I pass a number but we should be able to pass complex objects too 
*/

const count = rxSignal(0); // return union of function and Observable<int|undefined>

/**
"me" is a handle that let "count" register itself with "effect",
 I couldn't find another way to tell "effect" what is used inside,
 count will call the reference passed with "me" and let the "effect" successfully subscribe to it, 
 since the whole callback inside effect called at once, the effect can easily subscribe to it's usages
 this solution will solve the sync nature of signals too
 on each emit of "count", the "effect" will be called the same way that a signal works
*/
const unsub1= effect(me=>{
    console.log(count(me));
});

count.set(1); // effect will be called

// still it's an observable, we can subscribe and do other stuff with it 
const unsub2= count.subscribe(console.log); 

unsub1.destroy(); // it's calling unsubscribe early before component get destroyed
unsub2.unsubscribe(); // works as it should 

console.log(count()) // if no "me" is provided this will work like this.state.get('count');

effect(me=>{
    //I didn't test this, but just an idea, we can have some useful channel that let each  signal register itself with
     console.log(count(me.distinct)); 
});

count.set(1); //nothing will happen 

/** this will return a rxjs map of the callback
 basically, we check the type of me and based on that we call a function on me which is returning the actual value received 
 unfortunately, the solution that I ended up with pipe chaining wasn't elegant and I settled with having using only map (computed)
 so I don't post it here but I hope  that someone can find a way for it to pass multiple function like delay, throttle 
 I have some idea that maybe I share later 
*/
count double = computed(me => count(me)*2) 


effect(me=> {
    console.log(double(me));
});




```

**Comment by hoebbelsB** [comment]
@robert-king, thx for your valuable inputs :) I will try to cover all of your points!

> looks good, do we still need to provide RxState, RxEffects and RxAction factory? not sure how weak-refs work, i guess it's nice to be able to provide state at different injector levels depending on how shared the state is between components.

Technically you can provide just a single `RxState` instance at root level and share it across the entire application, no one is stopping you from doing that. This RFC should not change that behavior at all.

> Also the RxActionFactory isn't mentioned, would it be beneficial to have some of these signals convenience methods added to ui.actions?

Yep, the RFC is only targeting the `RxState` service. We will provide more RFCs for other packages / components of RxAngular. Thanks for the good point, we will think about if and how to integrate signals into the RxActionFactory.

> Also perhaps there's a way to configure RxAngular to default to signals or observables on a per-component bases, that way i could get compiler errors everywhere were i should change an observable to a signal, but i can also have backwards compatibility if i want. (It looks like you're going for backwards compatibility, although it'd be nice to be able to break things purposefully so that we can update our code to preferred methodology).

We do not want to drop the rxjs APIs of `RxState`, but instead deliver a single state management service that is versatile enough to treat signals and observables in one spot. This way the consumer always can decide which way to go. Signals are not a replacement for rxjs, just an additional reactive primitive. 
That being said, it is possible to introduce a single service that only provides the signal APIs. But it won't be a replacement for RxState, more like the tiny little brother which can be used in parallel.

**Comment by hoebbelsB** [comment]
hey @mhamri , thanks for your thoughts!

> I downvote this

What exactly? Our RFC or signals in general or an answer from someone else?

> 1- signal in angular is in a discovery state, which means it may work or it may not work. and it is depending on the team's findings after converting big projects to signals and if everything is still working as it should it's safe to build some API around that. I posted other places and mentioned it to @BioPhoton before. Signal has a memory issue and WeakRef is just one shot at it to see if that gets fixed or not.

True, it's in discovery state - same as we are right now :). But it will land in angular regardless. We are not here to "fix" bugs in the framework, but to discuss how to make the best out of the new APIs that will be introduced. Ofc the RFC / implementation of RxAngular will have to adjust if internal APIs are changing. 
The current draft is simply based on what is available for us now.

> 2- solids' signal is already working out of the box with angular; why we should re-implement it?

That's a question for the angular team and you might want to watch [this video](https://www.youtube.com/watch?v=HstDoVQeP9g) where the creator of solidjs is talking to pawel and alex about signals :).

Again, our draft is not about how to solve/implement/fix angular signals, but to integrate them.

> 3- preview version of angular is coming; what value are we adding here?

What exactly is the question? The value of preview versions? Scroll up - there u go. We, as library authors, can play with it and think about integrations eagerly.
Or is the question about the value of signals in general? Where should I start?

* local change detection
* optional zone.js
* fine grained reactivity
* no `async` pipe anymore
* leaner learning curve (because rxjs won't be used that heavy anymore)
* .... i'm sure I didn't point out everything 

> 4- **and most importantly,** rx-angular/state is used in production and not everyone is following latest versions of angular, getting dependency on a preview feature stop companies to update to latest version of rx-angular/state to get bug fixes, new features. ...

We will harm the libraries success if we don't react to changes made to the framework. We are already working on RxAngular versions dedicated to specific angular versions. But it completely depends on the available open source manpower if we are actually able to maintain older versions. 
We will welcome you with open arms if you like to contribute and help us pulling it off :).

> In against what we have here, I suggest rx way of signals. I prefer being able to inject multiple RxState and be able to build multiple states around it

You can already create as much `RxState` instances as you want, we didn't change that?

> 
> Moreover, RxState can have a single-type version that still works with observables
> 
> I already created a prototype of this and worked, I park it here as a suggestion

:eyes: 

> explanation inside the comments
> 
> ```ts
> /**
>  to return an observable and at the same time has signal API I used this way
>  I'm not sure if attaching more items to an observable is a good idea or not, but for the purpose of 
>  maintaining signal API I just bind a function to an observable
>  here I pass a number but we should be able to pass complex objects too 
> */
> 
> const count = rxSignal(0); // return union of function and Observable<int|undefined>
> 
> /**
> "me" is a handle that let "count" register itself with "effect",
>  I couldn't find another way to tell "effect" what is used inside,
>  count will call the reference passed with "me" and let the "effect" successfully subscribe to it, 
>  since the whole callback inside effect called at once, the effect can easily subscribe to it's usages
>  this solution will solve the sync nature of signals too
>  on each emit of "count", the "effect" will be called the same way that a signal works
> */
> const unsub1= effect(me=>{
>     console.log(count(me));
> });
> 
> count.set(1); // effect will be called
> 
> // still it's an observable, we can subscribe and do other stuff with it 
> const unsub2= count.subscribe(console.log); 
> 
> unsub1.destroy(); // it's calling unsubscribe early before component get destroyed
> unsub2.unsubscribe(); // works as it should 
> 
> console.log(count()) // if no "me" is provided this will work like this.state.get('count');
> 
> effect(me=>{
>     //I didn't test this, but just an idea, we can have some useful channel that let each  signal register itself with
>      console.log(count(me.distinct)); 
> });
> 
> count.set(1); //nothing will happen 
> 
> /** this will return a rxjs map of the callback
>  basically, we check the type of me and based on that we call a function on me which is returning the actual value received 
>  unfortunately, the solution that I ended up with pipe chaining wasn't elegant and I settled with having using only map (computed)
>  so I don't post it here but I hope  that someone can find a way for it to pass multiple function like delay, throttle 
>  I have some idea that maybe I share later 
> */
> count double = computed(me => count(me)*2) 
> 
> 
> effect(me=> {
>     console.log(double(me));
> });
> ```

I actually like that combination of observables and signals. Let's see what other folks are saying about this.
Also, to be able to create a `state` out of a simple function call is a very nice feature. With the [`DestroyRef`](https://twitter.com/Enea_Jahollari/status/1630665586245242881) & the `inject` method being added to the framework it could actually work quite well.

@rx-angular/core-team wdyt?

**Comment by tobang** [comment]
Any news on this RFC?

**Comment by hoebbelsB** [comment]
> Any news on this RFC?

hey @tobang yes, we are still actively working on this. Our main effort went into specifying a new simple public API for actions, effects and state which will be combined with signals asap.

* https://github.com/rx-angular/rx-angular/pull/1589
* https://github.com/rx-angular/rx-angular/pull/1591
* https://github.com/rx-angular/rx-angular/pull/1585
